### ISSUE TITLE
Error in VS2019

### ISSUE THREAD

**Issue by jampha** [cycfi/elements#382]
There was an error in VS2019. 

2>skia.lib(gpu.GrDrawingManager.obj) : error LNK2001: unresolved external symbol __std_find_trivial_8
2>skia.lib(skia.SkSLVMCodeGenerator.obj) : error LNK2001: unresolved external symbol __std_max_element_8
2>skia.lib(skia.SkSLString.obj) : error LNK2001: unresolved external symbol __std_find_trivial_1
2>skia.lib(gpu.GrGLCaps.obj) : error LNK2001: unresolved external symbol __std_find_trivial_4

Is it possible to provide the source code of the dependent library so that we can compile it ourselves.

**Comment by djowel** [comment]
We are using VS2022 currently. Any chance you cal use that instead. The linker errors are related to the c++ library. Skia is the dependent library, specifically tag m106. I must warn you though. Skia is a PAIN to build.

@redtide Another addition to the dependency matrix of the pre-compiled headers: compiler version :-(

**Comment by redtide** [comment]
[OK](https://github.com/redtide/skia-build/blob/master/todo.md?plain=1#L14)

**Comment by djowel** [comment]
@redtide The issue here is that we can't practically have all the precompiled binaries for all platforms, compilers, processors. And the glaring question is, what do we do when someone has a particular setup that is not included in our binaries for download? Do we just ask them to run the script? And then what?

**Comment by redtide** [comment]
Until now other than Arm64 and x86_64 (32 is becoming old even for RPi and the like it seems) there was only a request for RiskV (in another project).
Compilers? GCC, Clang and MSVC seems enough to me, maybe MingW64? Platforms... when you cover Linux/BSD, macOS and Windows, later maybe Android, IOS and WebAssembly IDK what else. Just check the [current repo](https://github.com/JetBrains/skia-pack) at JetBrains. There is a limit (we have limits), else... yes, DIY.

That is also why I like the idea to use scripts instead embedding commands in CI, so people can use them locally to build. Maybe that way they can propose some PR to include some exception we don't cover? Though it seems already a lot of builds to support.

As side note, the current JetBrains repo seems to have removed the compiler option, so I guess they now only build using GCC on Linux and MSVC on Windows. If were me I would force to use Clang everywhere instead since it's the one where Skia gives the maximum performances. IDK on Windows but using Clang on Linux is not that big pain for a C++ based project.

WDYM with and then what? I guess they have to check first also that the Artist, Elements (maybe also Infra header only libraries?) code supports their setup. They can also maintain their own custom setup themselves using CI (both for Skia and Elements) if we can't support all the options in the world.
In that case CI can also build Elements, so they can download the static lib directly for their project.

**Comment by djowel** [comment]
> Compilers? GCC, Clang and MSVC seems enough to me

That is correct. But I should have mentioned: compiler versions. As you see in this issue, a prebuilt library for VS 2019 is not necessarily compatible with VS 2020 or VS 2023.  It's a combinatorial explosion.

**Comment by djowel** [comment]
Here's a thought. What do you think about not pre-building at all using CI, but instead having the build script for Skia run, on demand, at install time on each platform, via cmake. That would mean, having python and all the build dependencies of Skia. 

The pre-built thing is looking like a nasty rabbit hole.

**Comment by redtide** [comment]
I'm not sure I understand what you mean, run the build at install time?
Don't we need to build Skia before Elements and then link to it when Elements' build finishes? 
From my POV you can add another option, say `ELEMENTS_USE_SKIA_PREBUILD` set to false by default, and run a custom CMake command that calls the python scripts. So if people like me don't want to build Skia, can use prebuilt binaries and set the option true/on.

**Comment by djowel** [comment]
> I'm not sure I understand what you mean, run the build at install time?

Build Skia using cmake, instead of downloading prebuilt binaries.

**Comment by redtide** [comment]
> > Compilers? GCC, Clang and MSVC seems enough to me
> 
> That is correct. But I should have mentioned: compiler versions. As you see in this issue, a prebuilt library for VS 2019 is not necessarily compatible with VS 2020 or VS 2023. It's a combinatorial explosion.

OK, so if version is also an issue I'm starting wondering how JetBrains people have no issues with it, maybe users using it just use the same compiler and same version? Still the answer above could be a solution: want different setup/unsupported platform/whatever? BIY (build it yourself :D ), else use our setup and prebuilt bins.

> Build Skia using cmake, instead of downloading prebuilt binaries.

Ah OK, so it's "configuration time" (IDK if there are "standard names"), install time for me means when you install all the built artifacts in system.
Yes, the idea is that after CMake do its tests/detections when configuring, it checks for the prebuilt bins option and if set it starts download binaries, else do nothing (default) and builds Skia when build starts.

**Comment by redtide** [comment]
In brief, we could not add the option now, and add Skia to the configuration:
- add its dependencies in CMake
- rename the build.py command to configure.py and remove the ninja build command
- run checkout.py and configure.py during CMake configuration
- at build time, before Elements, call Skia' ninja build command

**Comment by djowel** [comment]
> Yes, the idea is that after CMake do its tests/detections when configuring, it checks for the prebuilt bins option and if set it starts download binaries, else ~do nothing~ downloads and prepare Skia requirements/tools (default) and builds Skia when build starts.

Sounds good to me. Basically, something like: we prepared prebuilt libraries for Skia for xxx-system and www-compler and zzz-processor. If we do not have a prebulit library for your target, the cmake build scipt will attempt to build Skia for you. In such cases, here are the requirements: this and that and so on... python etc...

**Comment by djowel** [comment]
Please use the master (cairo) branch for now. Skia needs a LOT of work to get right.

**Comment by djowel** [comment]
**Update:** this is addressed on the `develop` branch (the new unified Skia + Cairo line) — not yet on `master`, but worth recording how it's resolved.

**Root cause recap:** the `__std_find_trivial_*` / `__std_max_element_*` unresolved-symbol errors are MSVC's vectorized-STL symbols. They appear when a prebuilt `skia.lib` was built with a *newer* MSVC toolset than the one linking it (here, VS2019) — i.e. prebuilt binaries aren't portable across compiler versions, exactly the "combinatorial explosion" discussed above.

**What `develop` now does** — essentially the "use a prebuilt if we have a matching one, otherwise build it for you" approach proposed in this thread:

- Skia is managed via vcpkg and also distributed as prebuilt **bundles keyed by OS + architecture + compiler class**.
- At configure time, CMake checks compatibility (MSVC toolset major, Linux glibc version, macOS deployment target). If a compatible prebuilt exists, it's downloaded and linked — **no Skia compile**. If not, you get a clear message to build from source (the `source` CMake preset / vcpkg toolchain), which compiles Skia **with your own compiler**, so the `__std_*` mismatch cannot occur.

So you always end up with either a matching prebuilt or a clean from-source build — no more silently-mismatched `skia.lib`.

**On VS2019 specifically:** the current Skia (m148) + C++20 effectively require VS2022, so that's the recommended toolchain — where the prebuilt bundle matches out of the box and a source build also works.

This becomes the default once `develop` is merged to `master`; we can close then.
